### PR TITLE
handle interrupts

### PR DESCRIPTION
**Pull request by toastbucket**

### CODE REVIEW

File: src/cpu/mod.rs [hunk 1]
```diff
@@ -654,6 +655,11 @@ impl Cpu {
         self.pc = 0x0100;
         self.set_sp(0xfffe);
     }
+
+    pub fn trigger_interrupt(&mut self, interrupt: Interrupt) {
```
Comment (toastbucket) [inline src/cpu/mod.rs:659]: I think I’m also missing a disable interrupt step here

File: src/gameboy/mod.rs [hunk 2]
```diff
@@ -75,12 +76,29 @@ impl Gameboy {
...
+            let mut should_trigger = false;
+
+            {
+                let intc = &self.mmu.borrow().intc;
+                should_trigger = intc.should_trigger(interrupt);
+            }
+
+            if should_trigger {
+                println!("triggering {:?}", interrupt);
+                self.cpu.trigger_interrupt(interrupt);
```
Comment (toastbucket) [inline src/gameboy/mod.rs:97]: Or perhaps here is where we should disable the interrupt. Need to clear the flag also

File: src/cpu/mod.rs (hunk 1 above)
Comment (ricfehr3) [reply src/cpu/mod.rs:668]: poop

**Comment by toastbucket** [comment]
@ricfehr3 I think we're ready to approach this PR now

File: src/intc/mod.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    pub fn should_trigger(&self, interrupt: Interrupt) -> bool {
+        self.is_enabled(interrupt) && self.is_requested(interrupt)
+    }
+
+    pub fn is_enabled(&self, interrupt: Interrupt) -> bool {
+        ((self.enable >> (interrupt as u8)) & 1) == 1
+    }
+    pub fn is_requested(&self, interrupt: Interrupt) -> bool {
+        (self.flag >> (interrupt as u8)) & 1 == 1
```
Comment (ricfehr3) [inline src/intc/mod.rs:103]: real good shit here

File: src/cpu/mod.rs [hunk 4]
```diff
@@ -958,6 +958,15 @@ impl Cpu {
                 self.halted = true;
                 (pc + 1, 1)
             },
+            // TODO: find a way to test these
```
Comment (ricfehr3) [inline src/cpu/mod.rs:961]: replace this comment saying there are no unit tests for these

File: src/gameboy/mod.rs [hunk 5]
```diff
@@ -96,7 +96,12 @@ impl Gameboy {
         'running: loop {
             self.handle_sdl2_events()?;
             self.handle_interrupts();
-            self.cpu.step();
+
+            if !self.cpu.halted() {
+                self.cpu.step();
+            } else {
+                sleep(Duration::from_millis(10));
```
Comment (ricfehr3) [inline src/gameboy/mod.rs:115]: Leave a comment saying this may dispensary.

Comment (toastbucket) [reply src/gameboy/mod.rs:115]: image

File: src/shell.rs [hunk 6]
```diff
@@ -27,9 +28,17 @@ pub struct Cmd {
     pub args: Vec<String>,
 }
 
-pub struct Shell;
+pub struct Shell {
+    dump_mode: bool,
```
Comment (ricfehr3) [inline src/shell.rs:32]: rename to dookie_mode?

Comment (toastbucket) [reply src/shell.rs:32]: big dump

File: src/int_src.rs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+// ░░█░░░▐░░░░░░░░▄░█░░░░░
+// ░░░█▄░░▀▄░░░░▄▀▐░█░░░░░
+// ░░░█▐▀▀▀░▀▀▀▀░░▐░█░░░░░
+// ░░▐█▐▄░░▀░░░░░░▐░█▄▄░░
+// ░░░▀▀░▄TSM▄░░░▐▄▄▄▀░░░
+
+pub trait InterruptSource {
+    // checks if interrupt source has pending interrupt
+    // if true is returned, internal flag will be cleared
+    fn check_and_consume_int_req(&mut self) -> bool {
```
Comment (ricfehr3) [inline src/int_src.rs:17]: love the word "consume"

File: src/joypad/mod.rs [hunk 8]
```diff
@@ -57,11 +69,35 @@ impl Joypad {
...
             buttons: [false; 8],
+            int_req: false,
         }
     }
 
     pub fn update_button(&mut self, button: Button, state: bool) {
         let idx = button as usize;
+        let cur_state = self.buttons[idx];
+
+        // TODO: emulate switch bounce?
```
Comment (ricfehr3) [inline src/joypad/mod.rs:80]: I don't think we should try to emulated switch bounce. I think sdl+keyboard/controller driver will debounce inputs. Putting that back in would be difficult and unnecessary.

File: src/cpu/mod.rs [hunk 9]
```diff
@@ -1136,6 +1136,7 @@ impl Cpu {
 
         self.pc = 0x0100;
         self.set_sp(0xfffe);
+        self.cycles = 0;
```
Comment (ricfehr3) [inline src/cpu/mod.rs:1139]: Reset joypad on reset as well?

File: src/cpu/test.rs [hunk 10]
```diff
@@ -3161,3 +3161,49 @@ fn test_swap_mem() {
...
+#[test]
+fn test_trigger_interrupt() {
+    let mut cpu = Cpu::new(Rc::new(RefCell::new(Mmu::new())));
+
+    cpu.reset();
+    cpu.trigger_interrupt(Interrupt::VBLANK);
+    let stack_top = cpu.get_reg_16(Register16Bit::SP);
+    let old_pc = cpu.read_word(stack_top);
+    assert_eq!(old_pc, 0x0100);
+    assert_eq!(cpu.pc, 0x0040);
```
Comment (ricfehr3) [inline src/cpu/test.rs:3174]: replace 0x0040 with interrupt vector variable?

Comment (ricfehr3) [reply src/cpu/test.rs:3174]: nah it's chill

File: src/intc/mod.rs (hunk 3 above)
Comment (toastbucket) [reply src/intc/mod.rs:103]: :D

File: src/cpu/mod.rs (hunk 4 above)
Comment (toastbucket) [reply src/cpu/mod.rs:961]: done

File: src/gameboy/mod.rs (hunk 5 above)
Comment (toastbucket) [reply src/gameboy/mod.rs:115]: done

File: src/int_src.rs (hunk 7 above)
Comment (toastbucket) [reply src/int_src.rs:17]: love u baby

File: src/joypad/mod.rs (hunk 8 above)
Comment (toastbucket) [reply src/joypad/mod.rs:75]: Yes they debounce, but some games expect bouncing on there.  Not sure if any rely one it.  Testing will tell.

File: src/cpu/mod.rs (hunk 9 above)
Comment (toastbucket) [reply src/cpu/mod.rs:1139]: joypad wont be reset from the CPU.  Resetting will be triggered from the Gameboy perspective, so once we implement that we can implement joypad reset.

**Review by ricfehr3 — APPROVED** [review]
Wow, this is starting to get... epic..? Nice PR dude!

**Pull request merged**